swarm/src/handler/from_fn.rs

```diff
@@ -0,0 +1,349 @@
+    protocol: &'static str,
+    inbound_streams_limit: usize,
+    pending_dial_limit: usize,
+    on_new_inbound: impl Fn(NegotiatedSubstream, &mut TState) -> TInboundFuture + Send + 'static,
+    on_new_outbound: impl Fn(NegotiatedSubstream, &mut TState, TOutboundOpenInfo) -> TOutboundFuture
+        + Send
+        + 'static,
+) -> FromFn<TInbound, TOutbound, TOutboundOpenInfo, TState>
+where
+    TInboundFuture: Future<Output = TInbound> + Send + 'static,
+    TOutboundFuture: Future<Output = TOutbound> + Send + 'static,
+    TState: Default,
+{
+    FromFn {
+        protocol,
+        inbound_streams_limit,
+        pending_dial_limit,
+        inbound_streams: FuturesUnordered::default(),
+        outbound_streams: FuturesUnordered::default(),
+        on_new_inbound: Box::new(move |stream, state| on_new_inbound(stream, state).boxed()),
+        on_new_outbound: Box::new(move |stream, state, info| {
+            on_new_outbound(stream, state, info).boxed()
+        }),
+        idle_waker: None,
+        pending_dials: VecDeque::default(),
+        failed_open: VecDeque::default(),
+        state: TState::default(),
+    }
+}
+
+#[derive(Debug)]
+pub enum OutEvent<I, O, OpenInfo> {
+    InboundFinished(I),
+    OutboundFinished(O),
+    FailedToOpen(OpenError<OpenInfo>),
+}
+
+#[derive(Debug)]
+pub enum OpenError<OpenInfo> {
+    Timeout(OpenInfo),
+    LimitExceeded(OpenInfo),
+    NegotiationFailed(OpenInfo, NegotiationError),
+}
+
+impl<OpenInfo> fmt::Display for OpenError<OpenInfo> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            OpenError::Timeout(_) => write!(f, "opening new substream timed out"),
+            OpenError::LimitExceeded(_) => write!(f, "limit for pending dials exceeded"),
+            OpenError::NegotiationFailed(_, _) => Ok(()), // Don't print anything to avoid double printing of error.
+        }
+    }
+}
+
+impl<OpenInfo> Error for OpenError<OpenInfo>
+where
+    OpenInfo: fmt::Debug,
+{
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            OpenError::Timeout(_) => None,
+            OpenError::LimitExceeded(_) => None,
+            OpenError::NegotiationFailed(_, source) => Some(source),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum InEvent<TState, TOutboundOpenInfo> {
+    UpdateState(TState),
+    NewOutbound(TOutboundOpenInfo),
+}
+
+pub struct FromFn<TInbound, TOutbound, TOutboundInfo, TState> {
+    protocol: &'static str,
+
+    inbound_streams: FuturesUnordered<BoxFuture<'static, TInbound>>,
+    outbound_streams: FuturesUnordered<BoxFuture<'static, TOutbound>>,
+
+    on_new_inbound:
+        Box<dyn Fn(NegotiatedSubstream, &mut TState) -> BoxFuture<'static, TInbound> + Send>,
+    on_new_outbound: Box<
+        dyn Fn(NegotiatedSubstream, &mut TState, TOutboundInfo) -> BoxFuture<'static, TOutbound>
+            + Send,
+    >,
```

Comment:
If we use a separate `FromFnProto` and implement `IntoConnectionHandler`, we would know the `PeerId` and `ConnectedPoint` of the remote and make it an parameter of the user's function. 
So it would then be `Fn(NegociatedSubstream, &mut TState, PeerId, ConnectedPoint)`.
Use case would be e.g. if we have different upgrades depending on whether a connection is relayed not, or depending on the PeerId, etc..

Reply:
That sounds like a good addition!

Reply:
I think this will allow for a really neat implementation of the identify protocol because we just have the observed address available in here :)